docs/proposals/operator-managed-infra.md

```diff
@@ -0,0 +1,128 @@
+# Operator Managed Infra and multi namespace
+Design doc: Not a feature yet
+
+Instead of managing all the infrastructure through terraform, the
+operator itself should be responsible for managing the Workload Identity
+bindings between the K8s Service Account and the Cloud Service Account/IAM role.
+
+Doing this, will make it possible to use Substratus resources in multiple
+namespaces without having to use another tool to configure Cloud identity/role mappings.
+
+Out of scope:
+* Managing the cluster configuration, node pools, dependent workpads/daemonsets, the bucket, image registry, the `sci-${provider}` principal, and its permissions.
+
+
+## Why?
+* Simplified install: Allow install to use `gcloud/aws` or `eksctl` to create initial K8s cluster + nodepools
+* Make install onto existing K8s cluster straight forward and remove
+  complexity of using Terraform
+* Improved UX, all an end-user needs to do is install our manifests and ensure GCP service account has enough permissions
+* Enterprises generally have very specific requirements about how they create their K8s Clusters and nodepools.
+* Allow Substratus to work in a multi namespace environment. Currently namespace is hardcoded in workload identity settings in the terraform
+
+## How?
+
+<img src="../diagrams/identity-global.excalidraw.png"></img>
+
+### Installation
+**Installation Flow** will be responsible for the following:
+* Create a K8s cluster with nodes/GPUs, this step is optional if end-user already has a K8s cluster (eksctl or gcloud)
+* Create Bucket, this step is optional if end-user already has a bucket
+* Create Image Registry, this step is optional if end-user already has an image registry
+* REQUIRED: Create Google Service Account or AWS IAM Role that has resource level permissions to:
+  * read and write a specific bucket
+  * pull and push to a specific Image Registry (GAR/ECR)
+  * (GCP) set IAM policy on the service Account itself to be able to manage workload identity bindings to KSAs. This can be done by assigning the service account to be a ServiceAccountAdmin to itself
+  * (AWS) UpdateAssumeRolePolicy on the AWS Role used by Substratus (required for multi namespace support)
+* REQUIRED: Apply Substratus manifests and configure them to use correct bucket and image registry
+
+### Controller
+Controller will be responsible for the following:
+* Creating the K8s ServiceAccount (already done today in `enforceServiceAccount`), in addition should now also call `sci-${CLOUD}:BindKSAToIAMPrincipal` to ensure the new KSA has access to the AWS IAM Role/GCP Service Account
+
+* (GCP only) example set annotation for Google Service Account (already done today) AND call `sci-gcp:BindKSAToIAMPrincipal` to update IAM policy on the Service Account so it can use the Google Service Account. For example:
```

Comment:
I think the call would just be `sci.BindKSAToIAMPrincipal` across all providers.

Should we separate out the GRPC services for adding and removing bindings? That seems implied in the decision to use `BindKSAToIAMPrincipal`. I was using `UpdateIamPrincipal` because I imagined an `operation: string` argument in the function signature with valid values of either `"add"` or `"remove"` so we'd be able to reuse the same service for both use cases. Let's either decide to:

1. consolidate into a single GRPC service or
2. decide on both add and removal service names (they're both needed) and add a detail on how that would look, what would trigger it, etc.

Reply:
I think we should never remove a K8s SA from an IAM principal. So vote for leaving that out of the proposal until we later see a need for it? When should we remove a K8s SA binding from IAM principal? I think it's better to leave those intact until the IAM principal itself is removed.

Reply:
> When should we remove a K8s SA binding from IAM principal?

If a namespace is deleted

Reply:
I'm ore concerned about the risk where an end-user has 2 clusters and in 1 cluster we remove the binding which causes his other cluster to also lose access unexpectedly. Hence I think we should just leave them in-place. Does AWS do cluster level bindings? If so that would be a good argument for implementing both add and remove. In GCP remove simply wouldn't do anything.

Reply:
Prefer to table cleanup for another day. Lets capture the concerns in an issue.

Reply:
Captured here: https://github.com/substratusai/substratus/issues/179